### PR TITLE
Migrate dapper to ef core

### PR DESCRIPTION
**Pull request by antonAce**

### CODE REVIEW

File: Src/Campus.Infrastructure.Data.EntityFrameworkCore/Repositories/AppUserRepository.cs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        private readonly CampusContext _context;
+        
+        public AppUserRepository(CampusContext context)
+        {
+            _context = context;
+        }
+        
+        public async Task CreateAppUserAsync(AppUser appUser)
+        {
+            await _context.Users.AddRangeAsync(appUser);
```
Comment (profileEnumerable) [inline Src/Campus.Infrastructure.Data.EntityFrameworkCore/Repositories/AppUserRepository.cs:21]: Do we need to use AddRangeAsync here ?

File: Src/Campus.Infrastructure.Business/Services/ProjectService.cs [hunk 2]
```diff
@@ -129,24 +125,13 @@ public async Task<int> AddTaskToProject(int id, TaskContentModelDto model)
...
-            {
-                throw new ApplicationException($"Project with specified id doesn't exist");
-            }
-
-            return affectedRows;
+            await _unitOfWork.CommitAsync();
         }
 
-        public async Task<int> DeleteProject(int userId, int projectId)
+        public async Task DeleteProject(int userId, int projectId)
```
Comment (antonAce) [inline Src/Campus.Infrastructure.Business/Services/ProjectService.cs:131]: Remove userId field

File: Src/Campus.Infrastructure.Business/Services/TaskService.cs [hunk 3]
```diff
@@ -9,16 +9,18 @@ namespace Campus.Infrastructure.Business.Services
...
         private readonly ITaskRepository _taskRepository;
 
-        public TaskService(ITaskRepository taskRepository)
+        public TaskService(IUnitOfWork unitOfWork, ITaskRepository taskRepository)
         {
+            _unitOfWork = unitOfWork;
             _taskRepository = taskRepository;
         }
 
         public async Task<TaskModelDto> GetTaskById(int userId, int taskId)
```
Comment (antonAce) [inline Src/Campus.Infrastructure.Business/Services/TaskService.cs:21]: Remove userId field

File: Src/Campus.Infrastructure.Data.EntityFrameworkCore/Configs/Mapping/ProjectMaps.cs [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+            builder.HasOne(p => p.Status)
+                .WithMany(s => s.Projects)
+                .HasForeignKey(p => p.StatusId)
+                .OnDelete(DeleteBehavior.Restrict);
+
+            builder.HasOne(p => p.User)
+                .WithMany(u => u.Projects)
+                .HasForeignKey(p => p.UserId)
+                .OnDelete(DeleteBehavior.Restrict);
```
Comment (antonAce) [inline Src/Campus.Infrastructure.Data.EntityFrameworkCore/Configs/Mapping/ProjectMaps.cs:28]: Use Cascade deletion here

File: Src/Campus.Infrastructure.Data.EntityFrameworkCore/Configs/Mapping/UserTaskMaps.cs [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+            builder.Property(ut => ut.ProjectTag)
+                .IsRequired().HasMaxLength(100);
+
+            builder.Property(ut => ut.Deadline)
+                .HasColumnType("datetime");
+
+            builder.HasOne(ut => ut.Project)
+                .WithMany(p => p.Tasks)
+                .HasForeignKey(ut => ut.ProjectId)
+                .OnDelete(DeleteBehavior.Restrict);
```
Comment (antonAce) [inline Src/Campus.Infrastructure.Data.EntityFrameworkCore/Configs/Mapping/UserTaskMaps.cs:29]: Use cascade deletion here

File: Src/Campus.Infrastructure.Data.EntityFrameworkCore/Dependencies/EntityFrameworkProvider.cs [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+    public static class EntityFrameworkProvider
+    {
+        public static void AddSqlServerStorage(this IServiceCollection services, string connectionString)
+        {
+            services.AddDbContext<CampusContext>(options => options.UseSqlServer(connectionString));
+            services.AddScoped<IUnitOfWork, UnitOfWork>();
+            services.AddScoped<IAppUserRepository, AppUserRepository>();
+            services.AddScoped<IProjectRepository, ProjectRepository>();
+            services.AddScoped<ITaskRepository, TaskRepository>();
+        }
```
Comment (antonAce) [inline Src/Campus.Infrastructure.Data.EntityFrameworkCore/Dependencies/EntityFrameworkProvider.cs:20]: Add PostgreSQL provider

File: Src/Campus.Infrastructure.Data.EntityFrameworkCore/Repositories/ProjectRepository.cs [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+        private readonly CampusContext _context;
+
+        public ProjectRepository(CampusContext context)
+        {
+            _context = context;
+        }
+
+        public async Task<IEnumerable<Project>> GetProjectsListing(int userId, int offset, int limit)
+        {
+            return await Task.Run(() => _context.Users.Include("Projects")
```
Comment (antonAce) [inline Src/Campus.Infrastructure.Data.EntityFrameworkCore/Repositories/ProjectRepository.cs:22]: Use explicit loading strategy

File: Src/Campus.Infrastructure.Data.EntityFrameworkCore/Repositories/ProjectRepository.cs [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+        public async Task EditProject(Project project)
+        {
+            await Task.Run(() => _context.Projects.Update(project));
+        }
+
+        public async Task<IEnumerable<UserTask>> GetProjectTasks(int userId, int projectId, int limit, int offset)
+        {
+            var project = await GetProjectById(projectId);
+
+            return project.Tasks.Skip(offset).Take(limit);
```
Comment (antonAce) [inline Src/Campus.Infrastructure.Data.EntityFrameworkCore/Repositories/ProjectRepository.cs:58]: Null check is missing here

File: Src/Campus.Infrastructure.Data.EntityFrameworkCore/Repositories/AppUserRepository.cs [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+        public async Task<AppUser> GetAppUserByLoginAsync(string login)
+        {
+            return await _context.Users.Where(u => u.Login == login)
+                                .FirstOrDefaultAsync();
+        }
+
+        public async Task DeleteAppUserByIdAsync(int id)
+        {
+            var user = await GetAppUserByIdAsync(id);
+            _context.Users.Remove(user);
```
Comment (antonAce) [inline Src/Campus.Infrastructure.Data.EntityFrameworkCore/Repositories/AppUserRepository.cs:39]: Add null check here

File: Src/Campus.Infrastructure.Data.EntityFrameworkCore/Repositories/TaskRepository.cs [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+        {
+            return await _context.Tasks
+                .FirstOrDefaultAsync(task => task.Id == taskId);
+        }
+
+        public async Task DeleteTask(int taskId)
+        {
+            var task = await GetTaskById(taskId);
+
+            _context.Tasks.Remove(task);
```
Comment (antonAce) [inline Src/Campus.Infrastructure.Data.EntityFrameworkCore/Repositories/TaskRepository.cs:28]: Add null check here

File: Src/Campus.Infrastructure.Data.EntityFrameworkCore/Repositories/UnitOfWork.cs [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        public async Task CommitAsync()
+        {
+            await _context.SaveChangesAsync();
+        }
+
+        public async Task RollbackAsync() => await Task.CompletedTask;
```
Comment (antonAce) [inline Src/Campus.Infrastructure.Data.EntityFrameworkCore/Repositories/UnitOfWork.cs:42]: Refactor this stub

File: Src/Campus.Master.API/Controllers/TaskController.cs [hunk 12]
```diff
@@ -70,15 +70,15 @@ public TaskController(ITaskService taskService, ILogger<TaskController> logger)
                     Description = "Description",
                     Priority = "Priority",
                     Tag = "Tag",
-                    Deadline = "00-00-0000"
+                    Deadline = DateTime.MinValue
```
Comment (antonAce) [inline Src/Campus.Master.API/Controllers/TaskController.cs:73]: Implement **Task** controller

File: Src/Campus.Infrastructure.Business/Services/ProjectService.cs [hunk 13]
```diff
@@ -45,19 +45,20 @@ public async Task<IEnumerable<ProjectModelDto>> GetSavedProjects(int userId, int
             return projectModels;
         }
 
-        public async Task<int> CreateProject(int userId, ProjectContentModelDto projectDto)
+        public async Task CreateProject(int userId, ProjectContentModelDto projectDto)
         {
-            return await _projectRepository.CreateNewProject(userId, new Project()
+            await _projectRepository.CreateNewProject(userId, new Project()
```
[the remark this answers is not in the document]
Comment (profileEnumerable) [reply Src/Campus.Infrastructure.Business/Services/ProjectService.cs:50]: ✔

File: Src/Campus.Infrastructure.Business/Services/ProjectService.cs [hunk 14]
```diff
@@ -73,9 +74,9 @@ public async Task<ProjectModelDto> GetProjectById(int userId, int projectId)
             };
         }
 
-        public async Task<int> EditProject(int userId, int id, ProjectContentModelDto projectContent)
+        public async Task EditProject(int userId, int id, ProjectContentModelDto projectContent)
```
[the remark this answers is not in the document]
Comment (profileEnumerable) [reply Src/Campus.Infrastructure.Business/Services/ProjectService.cs:77]: ✔

File: Src/Campus.Infrastructure.Business/Services/ProjectService.cs [hunk 15]
```diff
@@ -84,12 +85,7 @@ public async Task<int> EditProject(int userId, int id, ProjectContentModelDto pr
...
-            if (affectedRows == 0)
-            {
-                throw new ApplicationException($"Project with specified id doesn't exist");
-            }
-
-            return affectedRows;
+            await _unitOfWork.CommitAsync();
         }
 
         public async Task<IEnumerable<TaskModelDto>> GetProjectTasks(int userId, int id, int limit, int offset)
```
[the remark this answers is not in the document]
Comment (profileEnumerable) [reply Src/Campus.Infrastructure.Business/Services/ProjectService.cs:91]: ✔

File: Src/Campus.Infrastructure.Business/Services/ProjectService.cs (hunk 2 above)
Comment (profileEnumerable) [reply Src/Campus.Infrastructure.Business/Services/ProjectService.cs:131]: ✔

File: Src/Campus.Infrastructure.Business/Services/TaskService.cs (hunk 3 above)
Comment (profileEnumerable) [reply Src/Campus.Infrastructure.Business/Services/TaskService.cs:21]: ✔

File: Src/Campus.Infrastructure.Data.EntityFrameworkCore/Dependencies/EntityFrameworkProvider.cs (hunk 6 above)
Comment (profileEnumerable) [reply Src/Campus.Infrastructure.Data.EntityFrameworkCore/Dependencies/EntityFrameworkProvider.cs:20]: ✔

File: Src/Campus.Infrastructure.Data.EntityFrameworkCore/Repositories/ProjectRepository.cs (hunk 8 above)
Comment (profileEnumerable) [reply Src/Campus.Infrastructure.Data.EntityFrameworkCore/Repositories/ProjectRepository.cs:56]: ✔

File: Src/Campus.Infrastructure.Data.EntityFrameworkCore/Repositories/AppUserRepository.cs (hunk 9 above)
Comment (profileEnumerable) [reply Src/Campus.Infrastructure.Data.EntityFrameworkCore/Repositories/AppUserRepository.cs:39]: ✔

File: Src/Campus.Infrastructure.Data.EntityFrameworkCore/Repositories/TaskRepository.cs (hunk 10 above)
Comment (profileEnumerable) [reply Src/Campus.Infrastructure.Data.EntityFrameworkCore/Repositories/TaskRepository.cs:28]: ✔

File: Src/Campus.Infrastructure.Data.EntityFrameworkCore/Repositories/UnitOfWork.cs (hunk 11 above)
Comment (profileEnumerable) [reply Src/Campus.Infrastructure.Data.EntityFrameworkCore/Repositories/UnitOfWork.cs:42]: RollbackAsync was removed

File: Src/Campus.Master.API/Controllers/TaskController.cs (hunk 12 above)
Comment (profileEnumerable) [reply Src/Campus.Master.API/Controllers/TaskController.cs:73]: We need to define type of filter that can be applied

Comment (antonAce) [reply Src/Campus.Master.API/Controllers/TaskController.cs:73]: We can leave filtering endpoint empty but implement other

**Pull request merged**